internal/transport/transport_test.go

```diff
@@ -1248,6 +1254,57 @@ func (s) TestServerWithMisbehavedClient(t *testing.T) {
 	}
 }
 
+func (s) TestClientHonorsConnectContext(t *testing.T) {
+	// Create a server that will not send a preface.
+	lis, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("Error while listening: %v", err)
+	}
+	defer lis.Close()
+	go func() { // Launch the misbehaving server.
+		sconn, err := lis.Accept()
+		if err != nil {
+			t.Errorf("Error while accepting: %v", err)
+			return
+		}
+		defer sconn.Close()
+		if _, err := io.ReadFull(sconn, make([]byte, len(clientPreface))); err != nil {
+			t.Errorf("Error while reading client preface: %v", err)
+			return
+		}
+		sfr := http2.NewFramer(sconn, sconn)
+		// Do not write a settings frame, but read from the conn forever.
+		for {
+			_, err := sfr.ReadFrame()
+			if err != nil {
+				return
+			}
+		}
+	}()
+
+	// Test context cancelation.
+	timeBefore := time.Now()
+	connectCtx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	time.AfterFunc(100*time.Millisecond, cancel)
+
+	copts := ConnectOptions{ChannelzParentID: channelz.NewIdentifierForTesting(channelz.RefSubChannel, time.Now().Unix(), nil)}
+	_, err = NewClientTransport(connectCtx, context.Background(), resolver.Address{Addr: lis.Addr().String()}, copts, func(GoAwayReason) {}, func() {})
+	t.Logf("NewClientTransport() = _, %v", err)
+	if time.Now().Sub(timeBefore) > 2*time.Second {
```

Comment:
Where does this `2s` and `1.5s` come from. The `connectCtx` is canceled in `100ms`.

Reply:
I still don't understand why we want to be so generous with the time it takes to return from `NewClientTransport`. We cancel the context in `100ms`. Do we really have to give it `2s`?

Reply:
The tighter you make this check, the more likely you are to have a spurious failure caused by it taking too long to exit, which could be due to a hiccup on github actions / etc.  Ultimately, if we require that it is less than the deadline (3s) then we should be able to distinguish the reason it exited.  And actually, I'll extend the deadline and loosen this slightly since we're not concerned about how long it takes to work if it fails.

And, since there is no secondary cancelation mechanism in the second case, I just removed the time check entirely; the only failure is if no error is returned or if it never returns and times out because it's not honoring the context.